otherlibs/win32unix/dup.c

```diff
@@ -38,19 +38,24 @@ static SOCKET duplicate_socket(BOOL inherit, SOCKET oldsock)
 {
   WSAPROTOCOL_INFO info;
   SOCKET newsock;
+  DWORD flags = WSA_FLAG_OVERLAPPED;
+
   if (SOCKET_ERROR == WSADuplicateSocket(oldsock,
                                          GetCurrentProcessId(),
                                          &info)) {
     win32_maperr(WSAGetLastError());
     return INVALID_SOCKET;
   }
 
+  if (! inherit)
+    flags |= WSA_FLAG_NO_HANDLE_INHERIT;
+
   newsock = WSASocket(info.iAddressFamily, info.iSocketType, info.iProtocol,
-                      &info, 0, WSA_FLAG_OVERLAPPED);
+                      &info, 0, flags);
   if (INVALID_SOCKET == newsock)
     win32_maperr(WSAGetLastError());
-  else
-    win_set_inherit((HANDLE) newsock, inherit);
+  else if (! inherit)
+    win_set_inherit((HANDLE) newsock, FALSE);
```

Comment:
Do we really need to set `WAS_FLAG_NO_HANDLE_INHERIT` _and_ to call `win_set_inherit` afterwards?

Reply:
For old systems not supporting the flag, we still need to call `win_set_inherit`.

Reply:
Sure, but I would expect these old systems not supporting the flag to fail in `WSASocket`...

In other words, the code I was expecting is something like
```
  if WSA_FLAG_NO_HANDLE_INHERIT is supported:
    add it to flags if necessary
    call WSASocket
  else:
    call WSASocket
    call win_set_inherit
```

Reply:
> Sure, but I would expect these old systems not supporting the flag to fail in `WSASocket`...

You're probably right: I have tested calling `WSASocket` with high, undefined values of the flag parameter. I was expecting that it would ignore undefined values, but it fails as you expected with `WSAEPROTOTYPE`. I have tested it on Windows 10, I wasn't able to test on a older system (finding the VM images is tricky).
I was inspired by the [code in libuv](https://github.com/libuv/libuv/blob/0a47e4c7714c9d1cf62efef3efafd3fc3ad354c2/src/win/tcp.c#L1583-L1614) which also IIUC assumes `WSASocket` ignores the parameter, or it means that even with the flag, `WSASocket` could return an inheritable socket (and that would be very scary). You may have found a bug in libuv!

It's also possible that OCaml is build on a system defining `WSA_FLAG_NO_HANDLE_INHERIT` but run on an old system where winsock2 doesn't support this flag, or the opposite, OCaml may be build on an old system without the definition but run on a newer system.

What scenario do we want to support? I don't see a way to tell at _runtime_ whether the flag is supported or not, I think the best we can do is to try `WSASocket` with the flag and if it fails, retry without and use `SetHandleInformation`.

<details>
<summary>If we support old ⇋ new the wrapper may look like this, where we're forced to use both the flag and SetHandleInformation:</summary>

```c
SOCKET win32_socket(int domain, int type, int protocol,
                    LPWSAPROTOCOL_INFO info,
                    BOOL inherit)
{
  SOCKET s;
  DWORD flags = WSA_FLAG_OVERLAPPED;

#ifndef WSA_FLAG_NO_HANDLE_INHERIT
#define WSA_FLAG_NO_HANDLE_INHERIT 0x80
#endif

  if (! inherit)
    flags |= WSA_FLAG_NO_HANDLE_INHERIT;

  s = WSASocket(domain, type, protocol, info, 0, flags);
  if (s == INVALID_SOCKET) {
    if (! inherit && WSAGetLastError() == WSAEPROTOTYPE) {
      /* WSASocket probably doesn't suport WSA_FLAG_NO_HANDLE_INHERIT,
       * retry without. */
      flags &= ~(DWORD)WSA_FLAG_NO_HANDLE_INHERIT;
      s = WSASocket(domain, type, protocol, info, 0, flags);
      if (s == INVALID_SOCKET)
        goto err;
      win_set_inherit((HANDLE) s, FALSE);
      return s;
    }
    goto err;
  }

  if (! inherit)
    win_set_inherit((HANDLE) s, FALSE);

  return s;

err:
  win32_maperr(WSAGetLastError());
  return INVALID_SOCKET;
}
```
</details>

<details>
<summary>If however we support only new → old we could do:</summary>

```c
SOCKET win32_socket(int domain, int type, int protocol,
                    LPWSAPROTOCOL_INFO info,
                    BOOL inherit)
{
  SOCKET s;
  DWORD flags = WSA_FLAG_OVERLAPPED;

#ifdef WSA_FLAG_NO_HANDLE_INHERIT
  if (! inherit)
    flags |= WSA_FLAG_NO_HANDLE_INHERIT;
#endif

  s = WSASocket(domain, type, protocol, info, 0, flags);
  if (s == INVALID_SOCKET) {
#ifdef WSA_FLAG_NO_HANDLE_INHERIT
    if (! inherit && WSAGetLastError() == WSAEPROTOTYPE) {
      /* WSASocket probably doesn't suport WSA_FLAG_NO_HANDLE_INHERIT,
       * retry without. */
      flags &= ~(DWORD)WSA_FLAG_NO_HANDLE_INHERIT;
      s = WSASocket(domain, type, protocol, info, 0, flags);
      if (s == INVALID_SOCKET)
        goto err;
      win_set_inherit((HANDLE) s, FALSE);
      return s;
    }
#endif
    goto err;
  }

#ifndef WSA_FLAG_NO_HANDLE_INHERIT
  if (! inherit)
    win_set_inherit((HANDLE) s, FALSE);
#endif

  return s;

err:
  win32_maperr(WSAGetLastError());
  return INVALID_SOCKET;
}
```
</details>

I don't think we want to test at startup whether the runtime supports the flag, but that's a possibility too.